### PR TITLE
[BEAM-1116] Support for new Timer API in Flink runner (streaming)

### PR DESCRIPTION
**Pull request by JingsongLi**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by JingsongLi** [comment]
R: @aljoscha 
BTW, Is R reviewer? What does CC mean?

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/FlinkStateInternals.java [hunk 1]
```diff
@@ -75,7 +75,7 @@ public FlinkStateInternals(AbstractStateBackend flinkStateBackend, Coder<K> keyC
    * Returns the minimum over all watermark holds.
    */
   public Instant watermarkHold() {
-    long min = Long.MAX_VALUE;
+    long min = BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis();
```
Comment (JingsongLi) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/FlinkStateInternals.java]: BoundedWindow.TIMESTAMP_MAX_VALUE is the maximum value for any Beam timestamp. There is a ArithmeticException(The calculation caused an overflow) when use the Long.MAX_VALUE.

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 2]
```diff
@@ -446,6 +577,173 @@ public void restoreState(StreamTaskState state) throws Exception {
...
+
+    } else {
+      throw new RuntimeException("unexpected trigger: " + time);
+    }
+
+  }
+
+  protected void fireTimer(Tuple2<ByteBuffer, TimerInternals.TimerData> timer) {
+    TimerInternals.TimerData timerData = timer.f1;
+    pushbackDoFnRunner.onTimer(timerData.getTimerId(), GlobalWindow.INSTANCE,
```
Comment (JingsongLi) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: @kennknowles I use GlobalWindow refer to DirectRunner(ExecutorServiceParallelExecutor.fireTimers). Should I preserve BoundedWindow when user setTimer? I do not quite understand its meaning.

**Comment by JingsongLi** [comment]
follow-on work:
- support for new State and Timer API in Flink runner (batch)
- BEAM-867 Support set/delete of timers by ID in Flink runner
- Implement TestStream in Flink runner (to test processTimer and late data handle)

**Comment by aljoscha** [comment]
Hi @JingsongLi,
I love that you are putting so much work into the Flink runner but I think before changing this we need to tackle to other issues that I just created: 
 - https://issues.apache.org/jira/browse/BEAM-1393: Update Flink Runner to Flink 1.2.0
 - https://issues.apache.org/jira/browse/BEAM-1394: Use Flink InternalTimerService for TimerInternals

where the latter depends on the former. This will actually also simplify the implementation. Are you interested in working on those first?

For avoiding duplicate/redundant work it's good to mention that you are working on a Jira Issue and also to assign that issue to yourself. This way other people can now that you are working on something and won't do duplicate work.

Jira is also good for discussing implementation details.

Keep up the good work. 😄 👍 And let me know if you want to know anything and/or want to work on these other Jira issues I just created.

**Comment by JingsongLi** [comment]
@aljoscha Thank you for your advice. I will discuss more on Jira Issue. 
I am very glad to see that Flink 1.2.0 support InternalTimerService. This will simplify the implementation and be good for optimization.
I am very interested and want to work on these new feature.

**Comment by JingsongLi** [comment]
@aljoscha Is there a need to merge this PR? Or just close this PR and delete my fork?

**Comment by aljoscha** [comment]
@JingsongLi I think you can close the PR but keep your code/branch since the changes outside the operators will still be good once we move to Flink 1.2 and `InternalTimerService`.

**Pull request closed without merging**